sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ParameterAssignedTo.cs

```diff
@@ -72,13 +65,68 @@ protected override bool IsAssignmentToParameter(ISymbol symbol)
         {
             var parameterSymbol = symbol as IParameterSymbol;
             var result = parameterSymbol?.RefKind == RefKind.None;
-
             return result;
         }
 
+        protected override bool IsReadBefore(SemanticModel semanticModel, ISymbol symbol, AssignmentExpressionSyntax assignment)
+        {
+            return GetPreviousStatements(assignment).Union(new[] { assignment.Right })
+                .Select(s =>
+                {
+                    var collector = new ParameterAccessCollector(semanticModel, symbol);
+                    collector.Visit(s);
+                    return collector.IsRead;
+                })
+                .Aggregate((isRead, accumulator) => accumulator || isRead);
+        }
+
         protected override SyntaxNode GetAssignedNode(AssignmentExpressionSyntax assignment) => assignment.Left;
 
         protected sealed override GeneratedCodeRecognizer GeneratedCodeRecognizer => Helpers.CSharp.GeneratedCodeRecognizer.Instance;
+
+        /// <summary>
+        /// Returns all statements before the specified statement within the containing method.
+        /// This method recursively traverses all parent blocks of the provided statement.
+        /// </summary>
+        private static IEnumerable<SyntaxNode> GetPreviousStatements(SyntaxNode statement)
+        {
+            var previousStatements = statement.Parent.ChildNodes()
+                .OfType<StatementSyntax>()
+                .TakeWhile(x => x != statement)
+                .Reverse();
+
+            return statement.Parent is StatementSyntax parentStatement
+                ? previousStatements.Union(GetPreviousStatements(parentStatement))
+                : previousStatements;
+        }
+
+        private class ParameterAccessCollector : CSharpSyntaxWalker
+        {
+            public bool IsRead { get; private set; }
+            private readonly SemanticModel semanticModel;
+            private readonly ISymbol parameterSymbol;
+
+            public ParameterAccessCollector(SemanticModel semanticModel, ISymbol parameterSymbol)
+            {
+                this.semanticModel = semanticModel;
+                this.parameterSymbol = parameterSymbol;
+                IsRead = false;
+            }
+
+            public override void VisitIdentifierName(IdentifierNameSyntax node)
+            {
+                var nodeSymbol = semanticModel.GetSymbolInfo(node).Symbol;
+                IsRead |= (parameterSymbol == nodeSymbol && IsReadAccess(node));
+            }
+
+            private bool IsReadAccess(IdentifierNameSyntax node)
+            {
+                return !IsLeftSideOfAssignment(node);
+                bool IsLeftSideOfAssignment(SyntaxNode syntaxNode) =>

```

Comment:
[cosmetic] Please add a blank line before this inner method declaration.

Reply:
Actually you don't even need this sub-method.

Reply:
Because the function is affirmative, but inside it has a negation, I should either add a small comment or use a named method. I prefer the latter.

Reply:
@andrei-epure-sonarsource you can update the condition not to require the not.